src/test/scala/com/microsoft/hyperspace/index/IndexSignatureProviderTest.scala

```diff
@@ -0,0 +1,141 @@
+    val r1 = createSimplePlan(fileLength1, fileModificationTime1, filePath1)
+    val r2 = createLinearPlan(fileLength2, fileModificationTime2, filePath2)
+    assert(!createIndexSignature(r1).equals(createIndexSignature(r2)))
+  }
+
+  private def createSimplePlan(
+      length: Long,
+      modificationTime: Long,
+      path: Path): LogicalRelation = {
+    val c1 = AttributeReference("c1", StringType)()
+    val c2 = AttributeReference("c2", StringType)()
+
+    val tableSchema = schemaFromAttributes(c1, c2)
+    SignatureProviderTestUtils.createLogicalRelation(
+      spark,
+      Seq(createFileStatus(length, modificationTime, path)),
+      tableSchema)
+  }
+
+  private def createLinearPlan(length: Long, modificationTime: Long, path: Path): LogicalPlan = {
+    val t1c1 = AttributeReference("t1c1", StringType)()
+    val t1c2 = AttributeReference("t1c2", StringType)()
+    val t1c3 = AttributeReference("t1c3", IntegerType)()
+
+    val tableSchema = schemaFromAttributes(t1c1, t1c2, t1c3)
+    val scanNode = SignatureProviderTestUtils.createLogicalRelation(
+      spark,
+      Seq(createFileStatus(length, modificationTime, path)),
+      tableSchema)
+
+    val filterCondition = And(EqualTo(t1c1, Literal("ABC")), IsNotNull(t1c1))
+    val filterNode = Filter(filterCondition, scanNode)
+
+    Project(Seq(t1c2, t1c3), filterNode)
+  }
+
+  private def createComplexPlan: LogicalPlan = {
+    val t1c1 = AttributeReference("t1c1", StringType)()
+    val t1c2 = AttributeReference("t1c2", StringType)()
+    val t1c3 = AttributeReference("t1c3", IntegerType)()
+
+    val t1Schema = schemaFromAttributes(t1c1, t1c2, t1c3)
+    val r1 = SignatureProviderTestUtils.createLogicalRelation(
+      spark,
+      Seq(createFileStatus(fileLength1, fileModificationTime1, filePath1)),
+      t1Schema)
+
+    val t2c1 = AttributeReference("t2c1", IntegerType)()
+    val t2c2 = AttributeReference("t2c2", IntegerType)()
+
+    val t2Schema = schemaFromAttributes(t2c1, t2c2)
+    val r2 = SignatureProviderTestUtils.createLogicalRelation(
+      spark,
+      Seq(createFileStatus(fileLength2, fileModificationTime2, filePath2)),
+      t2Schema)
+
+    val joinCondition = EqualTo(t1c3, t2c2)
+    val joinNode = Join(r1, r2, JoinType("inner"), Some(joinCondition))
+
+    val filterCondition = And(EqualTo(t1c1, Literal("ABC")), IsNotNull(t1c1))
+    val filterNode = Filter(filterCondition, joinNode)
+
+    val projectNode = Project(Seq(t1c2, t1c3, t2c1), filterNode)
+
+    val grpExpression = Seq(t1c2, t1c3)
+    val aggExpression =
+      Seq(t1c2, t1c3, Alias(Count(Seq(t2c1)), "count(t2c1)")())
+
+    Aggregate(grpExpression, aggExpression, projectNode)
+  }
+
+  private def createIndexSignature(logicalPlan: LogicalPlan): String =
+    new IndexSignatureProvider().signature(logicalPlan) match {
+      case Some(s) => s
+      case None => throw HyperspaceException("Invalid plan for signature generation.")
```

Comment:
just `fail` here? unless you are testing the scenario where exception is thrown

Reply:
Done.